api/logs/v1/labels_enforcer.go

```diff
@@ -61,20 +62,56 @@ func WithEnforceAuthorizationLabels() func(http.Handler) http.Handler {
 
 const queryParam = "query"
 
-func enforceValues(mInfo AuthzResponseData, v url.Values) (values string, err error) {
-	if v.Get(queryParam) == "" {
-		return v.Encode(), nil
+func enforceValues(mInfo AuthzResponseData, u *url.URL) (values string, err error) {
+	switch {
+	case strings.HasSuffix(u.Path, "/values"):
+		return enforceValuesOnLabelValues(mInfo, u.Query())
+	default:
+		return enforceValuesOnQuery(mInfo, u.Query())
 	}
+}
 
-	lm := mInfo.Matchers
-	// Fix label matchers to include a non nil FastRegexMatcher for regex types.
-	for i, m := range lm {
-		nm, err := labels.NewMatcher(m.Type, m.Name, m.Value)
+func enforceValuesOnLabelValues(mInfo AuthzResponseData, v url.Values) (values string, err error) {
+	lm, err := initAuthzMatchers(mInfo.Matchers)
+	if err != nil {
+		return "", err
+	}
+
+	var expr logqlv2.Expr
+
+	q := v.Get(queryParam)
+	switch {
+	case q != "":
+		expr, err = logqlv2.ParseExpr(q)
 		if err != nil {
-			return "", fmt.Errorf("failed parsing label matcher: %w", err)
+			return "", fmt.Errorf("failed parsing LogQL expression: %w", err)
 		}
 
-		lm[i] = nm
+		switch le := expr.(type) {
+		case *logqlv2.LogQueryExpr:
+			matchers := combineLabelMatchers(le.Matchers(), lm)
+			le.SetMatchers(matchers)
+		default:
+			// Do nothing
+		}
+	default:
+		expr = &logqlv2.StreamMatcherExpr{}
+		expr.(*logqlv2.StreamMatcherExpr).SetMatchers(lm)
+	}
```

Comment:
An alternative way to avoid nested switch cases:
```golang
        var expr logqlv2.Expr = &logqlv2.StreamMatcherExpr{}
	expr.(*logqlv2.StreamMatcherExpr).SetMatchers(lm)

	q := v.Get(queryParam)
	if q != "" {
		expr, err = logqlv2.ParseExpr(q)
		if err != nil {
			return "", fmt.Errorf("failed parsing LogQL expression: %w", err)
		}

		switch le := expr.(type) {
		case *logqlv2.LogQueryExpr:
			matchers := combineLabelMatchers(le.Matchers(), lm)
			le.SetMatchers(matchers)
		default:
			// Do nothing
		}
	}
```